Recommendations

ederic-oytas opened the pull request:
# Pull Request Template

## Description

I added the /books/recommendations endpoint. This takes in three query parameters `titles`, `authors`, and `weights` which are sequences, whose elements are separated by `|---|`. These two parameters correspond with each other--titles[0] corresponds with authors[0] and weights[0], titles[1] corresponds with authors[1] and authors[1]. This endpoint returns a result in the same format as `/scan_book` and `/search`.

Weights should be calculated on the front-end. Maybe the formula can be: weight = (rating - 1) / 5 + 0.2 * favorite. So 0 star ratings would get 0 weight, 5-star ratings with no favorite would get 0.8 weight, and 5-star ratings with favorite would get 1.0 weight.

Major changes:
- Added `GET /books/recommendations` endpoint.
- Added `tags.py` which uses ChatGPT to generate tags given a set of titles and authors.
- Added an server API reference to `server/README.md`.

Minor changes:
- Fixed bug in `volume_info_to_book()` where the attributes of the returned dataclass instance weren't strings.
- Updated page_count and rating to be actual numbers and updated the samples

## Type of Change

- [x] Bugfix
- [X] New Feature
- [X] Documentation Update
- [ ] Other (please describe):

## Checklist

- [X] I have read the contributing guidelines.
- [X] I have followed the code style of this project.
- [X] I have added tests to cover my changes.
- [X] I have updated the documentation (if applicable).
- [X] My changes do not generate new warnings.

## Related Issue

#60 

## Additional Notes

<!-- Any other information or context relevant to the PR -->

LilAttacker commented on server/tabby_server/api/books.py:
```diff
@@ -175,3 +176,69 @@ def _get_result_dict(books: list[google_books.Book]) -> dict:
+            "message": "Unable to get tags from the given books."
+        }, HTTPStatus.BAD_REQUEST
+
+    # Get related books using list of tags
+    books = google_books.request_volumes_get(phrase=", ".join(tags_list))
```
Is this really a good way to go about this? I'm wondering if creating an individual request for each tag might return more relevant book information. Or perhaps searching by subject instead of general phrase might yield better results?

Just seems weird to me that we're searching for a book that would somehow have all the tags we want.

But if this works, then it works I guess.

LilAttacker replied:
Also, I'm pretty sure the space must go when making a request to Google Books. And since google_books.py doesn't format it, like replacing spaces with '+' or '%20' (which honestly seems like that may have been a mistake thinking about it), it should be done here pre-emptively.

ederic-oytas replied:
Yes, it works surprisingly well. At least Google Books doesn't return no books because of all the tags I give it. The purpose of using just one request is to make sure we don't hit the rate limit for Google Books calls.

> Also, I'm pretty sure the space must go when making a request to Google Books. And since google_books.py doesn't format it, like replacing spaces with '+' or '%20' (which honestly seems like that may have been a mistake thinking about it), it should be done here pre-emptively.

Good point. I don't think we've caught this error yet. It sort of reminds me of SQL injection. I'll try to follow up on this.

LilAttacker replied:
Google Books may return results, but are they actually relevant?

But fair enough on the concern about API calls, even if we're allotted a thousand a day, I could see us run out during presentation if we're too wasteful.

On that note, maybe we should add a way to disable the google books function to avoid hitting the cap that we could call from the server. Maybe make it return pre-gen results instead and we simply Steve Jobs the presentation and act like it's dynamic all the way through.

Probably not, but its a funny thought.

ederic-oytas replied:
Yes, the Google Books results are pretty relevant. And yeah, it would be kind of insane if an app started generating content the moment it runs out of money.

LilAttacker reviewed (changes requested):
After looking it over, it looks pretty good. I do have a few things I would like addressed before approval though.

--

First, we don't have a means to 'weigh' books. Like, if the user rated a fantasy book at 5 stars but a sci-fi books at 3, to my understanding if both were given to be used for recommendations then this would treat both as equally relevant when that's clearly not the case.

Are we just gonna rely on front-end to filter out those results for us by just giving us the highest rated books? What if those were the highest rated books?

I personally think a weight mechanic in the backend would be best, but I'd like to hear what you have to say.

--

Second, it seems to me that it's possible for recommendations to be "poisoned". Like if the books we received were mostly books of genre A and a little of genre B, I worry that genre B tags may be overrepresented.

Like if 6 tags of genre A were generated when it made up like 90% of the books but because they were all similar they didn't have enough unique tags to fill out the tag cap, so 4 tags of genre B were instead generated. This would cause an overrepresentation of genre B in recommendations.

Maybe that's what we want though, but I'd like to hear if this type of issue is actually possible or if I'm just worrying over nothing. (Or if it is possible, but attempting to fix it would be too annoying, I'd accept that as an answer too and approve).

--

Third, what was to occur if multiple books from the same author, or multiple books with the same title but different authors, were to be given to recommendations? I'm just wondering if we actually know and are prepared for it.

Otherwise, this doesn't need to be addressed really.

ederic-oytas commented:
1. Oh yes, I've forgotten about that feature. I'll need to implement it, although I'm unsure of whether ChatGPT will actually considers the weights. (It's kind of a blackbox with ChatGPT really)
2. I guess it could be "poisoned", although it would be expected that the reader would have more of one genre over another because they actually do prefer it. Anyway, I think this could be addressed more if we have the weights.
3. I think our system can handle it. It goes straight to ChatGPT, but I don't see it having any problems.

ederic-oytas commented:
By the way, I've made some changes:
- Updated `page_count` and `rating` to be actual numbers
- Updated the samples
- Added an server API reference.

LilAttacker reviewed (changes requested):
Alright, just updating my review.

I'll approve once a weight system has been added (or at least addressed further) and some kind of formatter has been added to the Google Books call functions.

Also, I saw the reference sheet, it looks good.

LilAttacker commented:
As an aside, how viable do you think it would be to create separate recommendation tabs?

Like having a section for "Recommended because you read this author" or "Recommended because you read this book", etc.

So that way we don't have a single monolith recommendations page, but one that's a bit more manageable and easier to traverse.

This probably falls under the "nice to have" category though.

ederic-oytas commented:
I've added a weighting system now. The /recommendations endpoint now receives a third array called `weights`. This contains numbers from 0 and 1. On ChatGPT, at least for extreme values, it works well when generating the tags.

These weights should be calculated on the front-end. Maybe the formula can be: weight = (rating - 1) / 5 + 0.2 * favorite. So 0 star ratings would get 0 weight, 5-star ratings with no favorite would get 0.8 weight, and 5-star ratings with favorite would get 1.0 weight.

ederic-oytas commented:
> As an aside, how viable do you think it would be to create separate recommendation tabs?
> 
> Like having a section for "Recommended because you read this author" or "Recommended because you read this book", etc.
> 
> So that way we don't have a single monolith recommendations page, but one that's a bit more manageable and easier to traverse.
> 
> This probably falls under the "nice to have" category though.

Actually, we might be able to implement that already on the front-end using the /search endpoint. Just find a book with a high rating by the user, and search for more books by that same author.

MrSnoodle01 commented:
Everything looks good. I'm really happy with the API reference in the server README, that should make everyone's life much easier when connecting the front and back end. Adding the final endpoint is great as well and looks good, I particularly like how the weights work.

David1000734 commented on server/tabby_server/services/tags.py:
```diff
@@ -0,0 +1,113 @@
+            messages=messages,  # type: ignore
+        )
+
+        # If no choices or response text given, try again
```
Probably not needed but a change to see if the what chatGPT actually responded with is a valid string of the format you specified would be good instead.

David1000734 reviewed (approved):
It looks good, comments were clear and I understand the intentions for each function

David1000734 commented on server/tabby_server/api/books.py:
```diff
@@ -175,3 +176,90 @@ def _get_result_dict(books: list[google_books.Book]) -> dict:
+            w = float(wstr)
+            if not (0.0 <= w <= 1.0):
+                return {
+                    "message": (f"Element {i} at 'weights' is not in [0, 1].")
```
I think it would be more accurate to say something like "weights is not within range (0.0 < x < 1.0) or something."
